test/CAPI/torch.c

```diff
@@ -0,0 +1,163 @@
+    fprintf(stderr, #TTT "I8 pos %d type: ", i);                               \
+    mlirTypePrint(torchMlirTorch##TTT##TypeGetType(TTT##I8, i), printToStderr, \
+                  NULL);                                                       \
+    fprintf(stderr, "\n");                                                     \
+  }
+  DEFINE_CHECK(Tuple)
+  DEFINE_CHECK(Union)
+#undef DEFINE_CHECK
+  // CHECK: TupleI8 isa Tuple: 1
+  // CHECK: TupleI8 NumTypes: 3
+  // CHECK: TupleI8 pos 0 type: i8
+  // CHECK: TupleI8 pos 1 type: f16
+  // CHECK: TupleI8 pos 2 type: f32
+  // CHECK: UnionI8 isa Union: 1
+  // CHECK: UnionI8 NumTypes: 3
+  // CHECK: UnionI8 pos 0 type: i8
+  // CHECK: UnionI8 pos 1 type: f16
+  // CHECK: UnionI8 pos 2 type: f32
+
+  int64_t sizes[3] = {1, 2, 3};
+  testTensor(ctx, 3, sizes, f32, "has-sizes-dtype");
+  // CHECK: NonValueTensorType has-sizes-dtype hasSizes: 1
+  // CHECK: NonValueTensorType has-sizes-dtype hasDtype: 1
+  // CHECK: NonValueTensorType has-sizes-dtype rank: 3
+  // CHECK: NonValueTensorType has-sizes-dtype pos 0 size: 1
+  // CHECK: NonValueTensorType has-sizes-dtype pos 1 size: 2
+  // CHECK: NonValueTensorType has-sizes-dtype pos 2 size: 3
+  // CHECK: NonValueTensorType has-sizes-dtype dtype: f32
+  // CHECK: ValueTensorType has-sizes-dtype hasSizes: 1
+  // CHECK: ValueTensorType has-sizes-dtype hasDtype: 1
+  // CHECK: ValueTensorType has-sizes-dtype rank: 3
+  // CHECK: ValueTensorType has-sizes-dtype pos 0 size: 1
+  // CHECK: ValueTensorType has-sizes-dtype pos 1 size: 2
+  // CHECK: ValueTensorType has-sizes-dtype pos 2 size: 3
+  // CHECK: ValueTensorType has-sizes-dtype dtype: f32
+
+  MlirType nullType = {NULL};
+  testTensor(ctx, 3, sizes, nullType, "has-sizes-no-dtype");
+  // CHECK: NonValueTensorType has-sizes-no-dtype hasSizes: 1
+  // CHECK: NonValueTensorType has-sizes-no-dtype hasDtype: 0
+  // CHECK: NonValueTensorType has-sizes-no-dtype rank: 3
+  // CHECK: NonValueTensorType has-sizes-no-dtype pos 0 size: 1
+  // CHECK: NonValueTensorType has-sizes-no-dtype pos 1 size: 2
+  // CHECK: NonValueTensorType has-sizes-no-dtype pos 2 size: 3
+  // CHECK: ValueTensorType has-sizes-no-dtype hasSizes: 1
+  // CHECK: ValueTensorType has-sizes-no-dtype hasDtype: 0
+  // CHECK: ValueTensorType has-sizes-no-dtype rank: 3
+  // CHECK: ValueTensorType has-sizes-no-dtype pos 0 size: 1
+  // CHECK: ValueTensorType has-sizes-no-dtype pos 1 size: 2
+  // CHECK: ValueTensorType has-sizes-no-dtype pos 2 size: 3
+  testTensor(ctx, -1, sizes, f32, "no-sizes-has-dtype");
+  // CHECK: NonValueTensorType no-sizes-has-dtype hasSizes: 0
+  // CHECK: NonValueTensorType no-sizes-has-dtype hasDtype: 1
+  // CHECK: NonValueTensorType no-sizes-has-dtype dtype: f32
+  // CHECK: ValueTensorType no-sizes-has-dtype hasSizes: 0
+  // CHECK: ValueTensorType no-sizes-has-dtype hasDtype: 1
+  // CHECK: ValueTensorType no-sizes-has-dtype dtype: f32
+
+  MlirType dict = torchMlirTorchDictTypeGet(f16, f32);
```

Comment:
This is actually not a valid dict type. The key and value type need to be torch dialect types, but we don't seem to verify that. Can you maybe add that check too? It should be just adding `let genVerifyDecl = 1;` to the td and implementing the method.

Reply:
Man this was an "excursion". 

Okay so firstly, indeed adding `genVerifyDecl = 1` is just a matter of implementing `LogicalResult DictType::verify` that checks key types and value types. But there are already [type predicates](https://github.com/llvm/torch-mlir/blob/7f95f4429795d80923a3a9c291ff5a4c77fdd180/include/torch-mlir/Dialect/Torch/IR/TorchTypes.td#L408) for this so I thought hmm I should try to use them somehow (save bytes in the final binary size) but it turns out that you can't do this ([upstream issue](https://github.com/llvm/llvm-project/issues/60015)).

Okay fine, code repetition it is, but then you gotta test it right? But adding lit/FileCheck tests, such as 

```mlir
// expected-error @+1 {{invalid torch.none for !torch.dict key type}}
func.func private @fail_dict() -> !torch.dict<none, tensor>
```

because the parser goes through the ["unchecked" factory](https://github.com/llvm/llvm-project/blob/main/mlir/docs/DefiningDialects/AttributesAndTypes.md#verification):

```
<unknown>:0: error: unexpected error: invalid '!torch.none' for !torch.dict key type
torch-mlir-opt:
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./bin/torch-mlir-opt -split-input-file -verify-diagnostics testdict.mlir
1.      MLIR Parser: custom op parser 'func.func'
2.      MLIR Parser: custom op parser 'torch.prim.DictConstruct'
 #0 0x0000564f22b5bf74 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int)
 #1 0x0000564f22b5c34a PrintStackTraceSignalHandler(void*)
 #2 0x0000564f22b59b68 llvm::sys::RunSignalHandlers()
 ...
#12 0x0000564f1ddd30db mlir::torch::Torch::DictType mlir::detail::StorageUserBase
#13 0x0000564f1dd498e1 mlir::torch::Torch::DictType::get(mlir::MLIRContext*, mlir::Type, mlir::Type)
#14 0x0000564f1dd4ad04 mlir::torch::Torch::DictType::parse(mlir::AsmParser&)
#15 0x0000564f1dd4889f generatedTypeParser(mlir::AsmParser&, ...
#16 0x0000564f1dd4b396
```

So therefore this has to be tested in driver/user code but I'm writing such a test anyway right? So I added/exposed 

```C
MLIR_CAPI_EXPORTED MlirType torchMlirTorchDictTypeGetChecked(
    MlirContext context, MlirType keyType, MlirType valueType);
```

which seems useful/reasonable enough but the implementation is annoyingly verbose:

```cpp
MlirType torchMlirTorchDictTypeGetChecked(MlirContext context, MlirType keyType,
                                          MlirType valueType) {
  auto unknownLoc = unwrap(mlirLocationUnknownGet(context));
  return wrap(Torch::DictType::getChecked(unknownLoc, unwrap(context),
                                          unwrap(keyType), unwrap(valueType)));
}
```

The reason you have to pass both the location *and* the context is because otherwise (for :shrug: knows why) you still hit the unchecked factory

```
// llvm-project/mlir/include/mlir/IR/StorageUniquerSupport.h

class StorageUserBase : public BaseT, public Traits<ConcreteT>... {
  ...
  static ConcreteT getChecked(const Location &loc, Args... args) {
    return ConcreteT::getChecked(getDefaultDiagnosticEmitFn(loc), args...);
  }
}

// build/tools/torch-mlir/include/torch-mlir/Dialect/Torch/IR/TorchTypes.cpp.inc

DictType DictType::get(::mlir::MLIRContext *context, ::mlir::Type keyType, ::mlir::Type valueType) {
  return Base::get(context, keyType, valueType);
}

DictType DictType::getChecked(::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError, ::mlir::MLIRContext *context, ::mlir::Type keyType, ::mlir::Type valueType) {
  return Base::getChecked(emitError, context, keyType, valueType);
}

DictType DictType::get(::mlir::Type keyType, ::mlir::Type valueType) {
  return Base::get(keyType.getContext(), keyType, valueType);
}

DictType DictType::getChecked(::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError, ::mlir::Type keyType, ::mlir::Type valueType) {
  return Base::get(keyType.getContext(), keyType, valueType);
}
```